string_indexing/suffix_array.py

```diff
@@ -63,6 +66,223 @@ def induced_sorting(text, n):
+          lambda x: SL[x] == Category.S, False, A, groups)
+    else:
+      raise Exception("fail")
+
+    return A.tab[1:]
+
+  def _sort_category(s, SL, category):
+    suff_by_dist, categ_substr = _prepare_substr_lists(
+        _bucket_array(s), SL, category)
+
+    # sort categ_substr
+    for i in range(1, len(suff_by_dist)):
+      groups = suff_by_dist[i].sorted_groups()
+      for g in groups if category == Category.S else reversed(groups):
+        to_split = collections.defaultdict(list)
+        for j in suff_by_dist[i].tab[g[0]:g[1]+1]:
+          x = j-i
+          g_id = categ_substr.group_id[categ_substr.rev[x]]
+          to_split[g_id].append(x)
+
+        for g_id in to_split:
+          vals = to_split[g_id]
+          if category == Category.S:
+            categ_substr.split_group_before(vals)
+            categ_substr.reverse_group(
+                categ_substr.group_id[categ_substr.rev[vals[0]]])
+          elif category == Category.L:
+            categ_substr.split_group_after(vals)
+            categ_substr.reverse_group(
+                categ_substr.group_id[categ_substr.rev[vals[0]]])
+
+    # map Ssubstr to ints
+    categ_substr.refresh_group_ids()
+    rec_s, m = [-1], { -1 : -1 }
+    for i in range(1, len(s)):
+      if SL[i] == category:
+        m[len(m)] = i
+        rec_s.append(categ_substr.group_id[categ_substr.rev[i]])
+
+    # recursive call and undo mapping
+    return [ m[i] for i in _small_large(rec_s) ]
+
+  def _move_to_group_end(vals, condition, front, A, groups):
+    for v in vals:
+      if not condition(v):
+        continue
+      idx = A.rev[v]
+      g_id = A.group_id[idx]
+      g = groups[g_id]
+
+      if front:
+        A.swap(idx, g[0])
+        groups[g_id] = (g[0]+1, g[1])
+      else:
+        A.swap(idx, g[1])
+        groups[g_id] = (g[0], g[1]-1)
+
+  def _prepare_substr_lists(A, SL, category):
+    categDist = _calc_category_dist(SL, category)
+    suff_by_dist = [ GroupedArray() for _ in range (max(categDist)+1) ]
+    category_substr = GroupedArray()
+
+    for g in A.sorted_groups():
+      to_add = []
+      for i in range(g[0], g[1]+1):
+        x = A.tab[i]
+        if SL[x] == category:
+          to_add.append(x)
+      category_substr.append_group(to_add)
+
+    for g in A.sorted_groups():
+      to_add = collections.defaultdict(list)
+      for i in range(g[0], g[1]+1):
+        x = A.tab[i]
+        d = categDist[x]
+        to_add[d].append(x)
+      for d in to_add:
+        suff_by_dist[d].append_group(to_add[d])
+
+    return (suff_by_dist, category_substr)
+
+  def _bucket_array(s):
+    buckets, A = [ [] for _ in range (len(s)) ], GroupedArray()
+    for i in range(1, len(s)):
+      buckets[s[i]].append(i)
+    for b in filter(lambda b: len(b) > 0, buckets):
+      A.append_group(b)
+    return A
+
+  def _categorize_sl(s):
+    SL, i = [ "." ], 1
+    while i < len(s)-1:
+      j = i+1
+      while s[i] == s[j]:
+        j += 1
```

Comment:
j = next(j for j in range(i + 1, len(s)) if s[i] != s[j]) if we are 100% sure that such element exists - and I guess it's certain, otherwise it'll throw Exception